### PR TITLE
Created Server View

### PR DESCRIPTION
**Pull request by ghost**
Created Server Views for handling GET,POST,PUT,DELETE.

### CODE REVIEW

File: moninag/server/models.py [hunk 1]
```diff
@@ -1,30 +1,156 @@
...
+        :param user_id: user id - optional parameter - used as filter
+        :param start: starting index of servers returned
+        :param end: las index of servers returned.
+        :return: QuerySet<Server>: QuerySet of servers.
+        """
+
+        filters = {}
+
+        if states:
+            filters['status__in'] = states
```
Comment (hutsulyk) [inline moninag/server/models.py]: Do you have `status` field? :upside_down_face:

File: moninag/server/urls.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+from django.views.decorators.csrf import csrf_exempt
+
+from . import views
+
+app_name = 'server'
+
+urlpatterns = [
+    url(r'^$', csrf_exempt(views.ServerView.as_view()), name='server'),
+    url(r'^(?P<server_id>\d+)/$', csrf_exempt(views.ServerView.as_view()), name='server'),
+]
```
Comment (hutsulyk) [inline moninag/server/urls.py:11]: Add end line.

Comment (ghost) [reply moninag/server/urls.py:11]: done

**Review by hutsulyk — APPROVED** [review]

File: moninag/server/models.py [hunk 3]
```diff
@@ -1,30 +1,156 @@
...
+        """Create and add server to database.
+
+        :param name: str - server name
+        :param address: str - server address
+        :param state: str - server state
+        :param user_id: id - user id
+        :return: Created server for success, None otherwise.
+        """
+
+        try:
```
Comment (lhalam) [inline moninag/server/models.py]: Remove user search unit,
The method receives the existing user_id

File: moninag/server/models.py [hunk 4]
```diff
@@ -1,30 +1,156 @@
...
+    :argument state: str - Server state ('NotSelected'/'Production','Staging')
+    :argument user: id - CustomUser id - user owner id
     """
     name = models.CharField(max_length=50)
     address = models.CharField(max_length=200)
-    state = models.CharField(max_length=100, choices=STATE_CHOICES, default=0)
+    state = models.CharField(max_length=20, choices=STATE_CHOICES, default='')
+    user = models.ForeignKey(CustomUser, default=1, on_delete=models.CASCADE)
+
+    @classmethod
```
Comment (lhalam) [inline moninag/server/models.py]: use @staticmetod

File: moninag/server/models.py [hunk 5]
```diff
@@ -1,30 +1,156 @@
...
+        except Exception as error:
+            return None
+
+        server = cls(name=name, address=address, state=state, user=user)
+        server.save()
+
+        return server
+
+    @classmethod
+    def update(cls, id, name, address, state, user_id):
```
Comment (lhalam) [inline moninag/server/models.py]: Are all parameters required?

File: moninag/server/models.py [hunk 6]
```diff
@@ -1,30 +1,156 @@
...
+        :param id: server id
+        :param name: server name
+        :param address: server address
+        :param state: server state
+        :param user_id: CustomUser id
+        :return: The return value. Updated server for success, None otherwise.
+        """
+
+        # Check if server exist allready
+        if cls.objects.filter(id=id).count() == 0:
```
Comment (lhalam) [inline moninag/server/models.py]: Remove user search unit,
The method receives the existing user_id and server_id

**Review by lhalam — COMMENTED** [review]
fix it all by analogy to comments

File: moninag/utils/validators.py [hunk 7]
```diff
@@ -10,3 +10,17 @@ def validate_dict(dictionary, requirements):
...
+
+    Args:
+        dictionary(dict): Dictionary to validate.
+        requirements(set): Set of required keys in dictionary.
+
+    Returns:
+        bool: The return value. True if valid, False otherwise.
+    """
+
+    return bool(set(dictionary.keys()).issubset(requirements))
```
Comment (hutsulyk) [inline moninag/utils/validators.py]: `issubset()` already return bool :slightly_smiling_face:

File: moninag/server/models.py [hunk 8]
```diff
@@ -1,30 +1,139 @@
...
+    def update(self, name=None, address=None, state=None):
+        """Update server data.
+
+        :param name: server name
+        :param address: server address
+        :param state: server state
+        :return: The return value. Updated server for success, None otherwise.
+        """
+
+        # Check if server exist allready
```
Comment (hutsulyk) [inline moninag/server/models.py]: Old comment :slightly_smiling_face:

File: moninag/server/models.py [hunk 9]
```diff
@@ -1,30 +1,139 @@
...
+        if name:
+            self.name = name
+        if address:
+            self.address = address
+        if state:
+            self.state = state
+        self.save()
+
+        return self
```
Comment (hutsulyk) [inline moninag/server/models.py]: Is it necessary to return self?

File: moninag/server/views.py [hunk 10]
```diff
@@ -1,2 +1,141 @@
...
+            # Get server with specified id
+            server = Server.get_by_id(server_id)
+            if server:
+                if server.user.id == request.user.id:
+                    json_response['response'] = server.to_dict()
+                    return JsonResponse(json_response, status=200)
+                else:
+                    return HttpResponse(code=403)
+            else:
+                HttpResponse(code=400)
```
Comment (hutsulyk) [inline moninag/server/views.py]: Did you miss `return`? :slightly_smiling_face:

File: moninag/server/views.py [hunk 11]
```diff
@@ -1,2 +1,140 @@
...
+            return JsonResponse(json_response, status=400)
+
+        server = Server.create(user=request.user, **server_dict)
+
+        if server:
+            json_response['response'] = server.to_dict()
+            return JsonResponse(json_response, status=201)
+
+        json_response['error'] = 'Server was not created.'
+        return JsonResponse(json_response, status=400)
```
Comment (hutsulyk) [inline moninag/server/views.py]: When `if server:` will be not true?

File: moninag/server/views.py [hunk 12]
```diff
@@ -1,2 +1,140 @@
...
+            json_response['error'] = 'Incorect JSON format.'
+            return JsonResponse(json_response, status=400)
+
+        server = Server.get_by_id(server_id)
+
+        if server:
+            if server.user.id == request.user.id:
+                server.update(**server_dict)
+                json_response['response'] = server.to_dict()
+                return JsonResponse(json_response, status=200)
```
Comment (hutsulyk) [inline moninag/server/views.py]: Return 403 `if server.user.id == request.user.id:` False

File: moninag/server/views.py [hunk 13]
```diff
@@ -1,2 +1,140 @@
...
+        :param server_id: int - server id
+        :return: HttpResponse: Status 200 for success, 400 otherwise.
+        """
+        server = Server.get_by_id(server_id)
+        if server.user.id == request.user.id:
+            count = server.delete()
+            if count:
+                return HttpResponse(status=200)
+        else:
+            return HttpResponse(status=400)
```
Comment (hutsulyk) [inline moninag/server/views.py]: status = 403 :slightly_smiling_face:

File: moninag/server/views.py [hunk 14]
```diff
@@ -1,2 +1,142 @@
...
+                    json_response['response'] = server.to_dict()
+                    return JsonResponse(json_response, status=200)
+                else:
+                    return HttpResponse(code=403)
+            else:
+                json_response['error'] = 'Server with specified id was not found.'
+                return JsonResponse(json_response, status=404)
+
+        # Get all servers
+        if request.user.id:
```
Comment (hutsulyk) [inline moninag/server/views.py]: `if request.user.is_authenticated:` :slightly_smiling_face:

File: moninag/server/views.py (hunk 12 above)
Comment (ghost) [reply moninag/server/views.py]: @mhumetskyi fixed

File: moninag/server/models.py [hunk 15]
```diff
@@ -1,30 +1,136 @@
...
+        except Exception as error:
+            return None
+
+        return server
+
+    @staticmethod
+    def get_by_user_id(user_id):
+        """
+        :param user_id:
+        :return: return Server
```
Comment (lhalam) [inline moninag/server/models.py]: `:param user_id:` type param: Int or user object?
`:return: return Server` Returns only one server or server list?

File: moninag/server/views.py (hunk 14 above)
Comment (lhalam) [reply moninag/server/views.py]: do not need

File: moninag/server/views.py [hunk 16]
```diff
@@ -1,2 +1,142 @@
...
+
+        json_response = {}
+
+        server_dict = json.loads(request.body.decode('utf-8'))
+
+        if not validate_dict(server_dict, REQUIREMENTS):
+            json_response['error'] = 'Incorect JSON format.'
+            return JsonResponse(json_response, status=400)
+
+        if request.user.id:
```
Comment (lhalam) [inline moninag/server/views.py]: No need to check
Update implementing

**Review by lhalam — APPROVED** [review]

**Pull request merged**